### DISCUSSION TITLE
Is there any more detailed documentation on encoding?

### DISCUSSION THREAD

**Discussion by holdxen** [capnproto/capnproto#2110] [Q&A]
Currently `capnproto-rust` is not thread-safe and cannot be used for multithreading. I want to implement a version that supports multithreading and thread safety. However, after reading https://capnproto.org/encoding.html, I still have only a vague understanding of the encoding format. Is there any more detailed documentation?

**Comment by kentonv** [comment]
No, there's no other documentation of the encoding. Feel free to ask questions about anything that's unclear.

**Reply by holdxen** [reply]
![image](https://github.com/user-attachments/assets/b45f97ff-e973-49a9-a3a6-6e0529a762a3)

Thanks for help.
B is signed, does that mean B can be negative (except for the empty struct)? But I don't see any check for segment out of bounds when B is negative in Rust code. Am I wrong?

**Reply by kentonv** [reply]
Yes, B can be negative. Objects can appear in any order within a segment, so a pointer in one object may have to point to an object appearing earlier in the segment, making the offset negative.

I don't know the Rust code, can you post a link to where you're looking?

**Reply by holdxen** [reply]
Thanks.

code screenshot:
![image](https://github.com/user-attachments/assets/8e29b300-d64e-4705-abfb-5aec85886cbf)

link:
https://github.com/capnproto/capnproto-rust/blob/e068cba1938c945aef151f8422cea03c5009e082/capnp/src/private/arena.rs#L105

**Reply by kentonv** [reply]
So, this part:

`((start_idx - this_start) as i64 + offset) as usize > this_size`

`((start_idx - this_start) as i64 + offset)` computes the pointed-to location relative to the start of the segment. If the pointer points before the start of the segment, this will be negative. Then, `as usize` coerces it to unsigned, which means negative numbers will wrap around and become very large positive numbers instead. So, `> this_size` will be true, so it'll be detected as out-of-bounds.

Maybe @dwrensha could confirm.

**Reply by holdxen** [reply]
You are right, I completely understand this code now. Thanks for help.
